precompiles/common/precompile.go

```diff
@@ -128,19 +139,28 @@ func (p Precompile) RunSetup(
 	// we need to consume the gas that was already used by the EVM
 	ctx.GasMeter().ConsumeGas(initialGas, "creating a new gas meter")
 
-	// add a snapshot of the current state before executing the precompile
-	// so that any errors during said execution are reverted correctly
-	if isTransaction(method.Name) {
-		if err := stateDB.AddPrecompileFn(p.Address(), multiStore, events); err != nil {
-			return sdk.Context{}, nil, nil, sdk.Gas(0), nil, err
-		}
-		// native balance changes should be added in Run by the precompile.
-	}
-
 	// return the error (set by HandleGasError) or nil
 	if err != nil {
 		return sdk.Context{}, nil, nil, sdk.Gas(0), nil, err
 	}
+
+	// dump every in-memory stateDB change to the in-memory cached context.
+	// note that no disk commitment happens here.
+	// previously, this was done at the top of this function and not here at
+	// the bottom. that caused issues in case something else errored out
+	// after the `CommitWithCacheCtx` call.
+	// for example, if the precompile was called with an unknown method ID,
+	// and stateDB.CommitWithCacheCtx() was called first, the cached context
+	// would store an incorrect state without a corresponding journal entry.
+	// in other words, this below function dumps the changes into `writeCache`,
+	// which would override the disk-commitment with incorrect values.
+	if isTx {
+		// only commit when we aren't read-only
+		if err := stateDB.CommitWithCacheCtx(); err != nil {
```

Comment:
good catch! for precompile `RunSetup`, we should either: 1. always make sure we add snapshot to journal entry in any case, or 2. only call `CommitWithCacheCtx` when we can ensure that the snapshot would be added to the journal entry. Is this principle correct? 

In this fix, we have taken the second approach: only call `CommitWithCacheCtx` to write dirty storage to cache ctx when we are able to add snptshot to journal entry, otherwise the whole `precompile.Run` would be like a no-op call(there is no storage write) so that no journal entry needed and we just let the revert pop up.

could we put the two function calls together for `isTx`(true) branch to make this more explicit if my understanding is correct?

Reply:
it's ok, because: the snapshot is the multistore within `cacheCtx` and it is taken before we call `CommitWithCacheCtx`, so when any error happens for either `precompie.RunSetup` or `precompile.Run`, as long as we have the correct journal entry(snapshot), the `cacheCtx` would be reverted to the state before we execute the precompile

Reply:
besides, after looking deeper into this, I think: 1. we should always call `CommitWithCacheCtx`, not only for method that is `isTx`, but also methods that only read, since we need to make sure the multi store that `precompile.Run` accesses includes all the state changes from previous calls, 2. we should always add journal entries for any case, so that the `cacheCtx` could be reverted to the state before precompile execution

Reply:
> > could we put the two function calls together for isTx(true) branch to make this more explicit if my understanding is correct?
> 
> Not without moving `isTx` downwards. If we call `CommitWithCacheCtx`, and then `method.Inputs.Unpack` errors out, we are left with different state inside the cached context. The journal entry does not revert this state. Hence, an unexpected value ends up being written to disk.

this situation is equal to `precompile.Run` errors out after calling `precompile.RunSetup`, which means a revert, so it would be ok if we have the correct journal entry